File: service-jobs/tests/RabbitMq.spec.ts
import * as assert from 'assert';
import {
    RabbitMqConnection,
    RabbitMqMessageReceiver,
    RabbitMqMessageSender,
} from '../src/controllers/RabbitMq';
import config from '../src/configs/server';

describe('will test rabbit mq helpers', function () {
    this.timeout(5000);

    const queue1 = 'tester';
    const queue2 = 'tester-2';
    const jobId = 'test-job-id';
    const corrId = 'test-corr-id';
    const jweToken = 'encrypted-jwt-token';
    const senderId = config.rabbitMq.serviceQueueName;

    const message = 'this is a test message';

    it('should connect and disconnect from rabbitmq', async () => {
        const connection = new RabbitMqConnection(queue1);
        assert(connection);
        assert.equal(connection.isConnected(), false);
        await connection.connect();
        assert(connection.isConnected());
        await connection.close();
        assert.equal(connection.isConnected(), false);
        assert.equal(connection.queueName, queue1);
    });

    it('should send a message with sender', async () => {
        const sender = new RabbitMqMessageSender(queue1);
        assert(sender);
        assert.equal(sender.isConnected(), false);
        await sender.sendJsonMessage(queue1, message, jobId, corrId, jweToken, false);
        assert(sender.isConnected());
        assert.equal(sender.getQueueName(), queue1);
        await sender.close();
        assert.equal(sender.isConnected(), false);
    });

    it('should send message from a to b and include header data in message body', async () => {
        const sender = new RabbitMqMessageSender(queue2);
        const receiver = new RabbitMqMessageReceiver(queue2);
        assert(sender);
        assert(receiver);
        assert.equal(receiver.isConnected(), false);
        assert.equal(sender.getQueueName(), queue2);
        assert.equal(receiver.getQueueName(), queue2);
        await sender.sendJsonMessage(queue2, message, jobId, corrId, jweToken, false);
        assert(sender.isConnected());

        await new Promise<void>(async (resolve, reject) => {
            await receiver.getMessagesOnQueue<string>((messageReceived, messageSource) => {
                assert.equal(messageReceived.from, senderId);
                assert.equal(messageReceived.to, queue2);
                assert.equal(messageReceived.jobId, jobId);
                assert.strictEqual(messageReceived.message, message);
                const today = new Date().toISOString().split('T').shift();
                const messageDate = messageReceived.time.split('T').shift();
                assert.equal(messageDate, today);
                const corrIdFromMessage = messageSource.properties.headers?.['x-correlation-id'];
                const jweTokenFromMessgae = messageSource.properties.headers?.authorization;
                assert.equal(corrIdFromMessage, corrId);
                assert.equal(jweTokenFromMessgae, `Bearer ${jweToken}`);

                receiver.acknowledgeMessageReceipt(messageSource);

                resolve();
            });
        });

        await sender.close();
        await receiver.close();

        assert.equal(sender.isConnected(), false);
        assert.equal(receiver.isConnected(), false);
    });
});

File: service_python_shared/service_python_shared/tests/rabbitmq_spec.py
import asyncio
from aio_pika import IncomingMessage
from datetime import datetime, timezone
import pytest

from service_python_shared.modules.rabbitmq import (
    RabbitMqConnection,
    RabbitMqMessageSender,
    RabbitMqMessageReceiver,
    RabbitMqMessage,
)
from service_python_shared.configs.config import config


QUEUE_1 = "tester"
QUEUE_2 = "tester-2"
JOB_ID = "test-job-id"
CORR_ID = "test-corr-id"
JWE_TOKEN = "encrypted-jwt-token"
SENDER_ID = config["rabbitmq"]["service_queue_name"]
TEST_MESSAGE = "this is a test message"


@pytest.mark.asyncio
@pytest.mark.timeout(5)
async def test_connect_and_disconnect():
    conn = RabbitMqConnection(QUEUE_1)
    assert conn
    assert conn.is_connected() is False
    await conn.connect()
    assert conn.is_connected() is True
    await conn.close()
    assert conn.is_connected() is False
    assert conn.queue_name == QUEUE_1


@pytest.mark.asyncio
@pytest.mark.timeout(5)
async def test_send_message_with_sender():
    sender = RabbitMqMessageSender(QUEUE_1)
    assert sender
    assert sender.is_connected() is False

    await sender.send_json_message(
        queue_name=QUEUE_1,
        message=TEST_MESSAGE,
        filepath="dummy-path",
        md5="dummy-md5",
        job_id=JOB_ID,
        corr_id=CORR_ID,
        jwe_token=JWE_TOKEN,
        persistent=False,
    )

    assert sender.is_connected() is True
    assert sender.get_queue_name() == QUEUE_1
    await sender.close()
    assert sender.is_connected() is False


@pytest.mark.asyncio
@pytest.mark.timeout(5)
async def test_send_message_from_a_to_b_and_check_headers():
    sender = RabbitMqMessageSender(QUEUE_2)
    receiver = RabbitMqMessageReceiver(QUEUE_2)

    assert sender and receiver
    assert receiver.is_connected() is False
    assert sender.get_queue_name() == QUEUE_2
    assert receiver.get_queue_name() == QUEUE_2

    await sender.send_json_message(
        queue_name=QUEUE_2,
        message=TEST_MESSAGE,
        filepath="dummy-path",
        md5="dummy-md5",
        job_id=JOB_ID,
        corr_id=CORR_ID,
        jwe_token=JWE_TOKEN,
        persistent=False,
    )
    assert sender.is_connected()

    # Wrap in a timeout so the test doesn't hang indefinitely
    async def receive_once():
        fut = asyncio.get_event_loop().create_future()

        async def handler(
            message_received: RabbitMqMessage[str], message_source: IncomingMessage
        ):
            assert message_received.from_ == SENDER_ID
            assert message_received.to == QUEUE_2
            assert message_received.jobId == JOB_ID
            assert message_received.message == TEST_MESSAGE
            today = datetime.now(timezone.utc).date().isoformat()
            message_date = message_received.time.split("T")[0]
            assert message_date == today

            corr_id_header = message_source.headers.get("x-correlation-id")
            auth_header = message_source.headers.get("authorization")

            assert corr_id_header == CORR_ID
            assert auth_header == f"Bearer {JWE_TOKEN}"

            await message_source.ack()
            fut.set_result(True)

        await receiver.get_messages_on_queue(handler)
        await asyncio.wait_for(fut, timeout=3)

    await receive_once()

    await sender.close()
    await receiver.close()

    assert sender.is_connected() is False
    assert receiver.is_connected() is False
